network/discovery/fork_listener.go

```diff
@@ -0,0 +1,160 @@
+// After the fork, returns only the result from the post-fork service.
+func (l *forkListener) AllNodes() []*enode.Node {
+	if l.netCfg.PastAlanFork() {
+		l.closePreForkListener()
+		return l.postForkListener.AllNodes()
+	}
+
+	enodes := l.postForkListener.AllNodes()
+	enodes = append(enodes, l.preForkListener.AllNodes()...)
+	return enodes
+}
+
+// Sends a ping in the post-fork service.
+// Before the fork, it also tries to ping with the pre-fork service in case of error.
+func (l *forkListener) Ping(node *enode.Node) error {
+	if l.netCfg.PastAlanFork() {
+		l.closePreForkListener()
+		return l.postForkListener.Ping(node)
+	}
+
+	err := l.postForkListener.Ping(node)
+	if err != nil {
+		return l.preForkListener.Ping(node)
+	}
+	return nil
+}
+
+// Returns the LocalNode using the post-fork listener.
+// Both pre and post-fork listeners should have the same LocalNode.
+func (l *forkListener) LocalNode() *enode.LocalNode {
+	if l.netCfg.PastAlanFork() {
+		l.closePreForkListener()
+		return l.postForkListener.LocalNode()
+	}
+	return l.postForkListener.LocalNode()
+}
+
+// Closes both listeners
+func (l *forkListener) Close() {
+	l.closePreForkListener()
+	l.postForkListener.Close()
+}
+
+// closePreForkListener ensures preForkListener is closed once
+func (l *forkListener) closePreForkListener() {
+	l.closeOnce.Do(func() {
+		l.preForkListener.Close()
+	})
+}
+
+// preAndPostForkIterator is an Iterator wrapper for enode.Iterator
+// It gives preference to the postForkIterator, starting by it
+// After the postForkIterator ends, it starts using the preForkIterator
+type preAndPostForkIterator struct {
+	preForkIterator             enode.Iterator
+	postForkIterator            enode.Iterator
+	hasChangedToPreForkIterator bool
+	mutex                       sync.Mutex
+}
+
+func NewPreAndPostForkIterator(preForkIterator enode.Iterator, postForkIterator enode.Iterator) enode.Iterator {
+	return &preAndPostForkIterator{
+		preForkIterator:             preForkIterator,
+		postForkIterator:            postForkIterator,
+		hasChangedToPreForkIterator: false,
+		mutex:                       sync.Mutex{},
+	}
+}
+
+func (i *preAndPostForkIterator) Close() {
+	i.mutex.Lock()
+	defer i.mutex.Unlock()
+
+	i.preForkIterator.Close()
+	i.postForkIterator.Close()
+}
+
+func (i *preAndPostForkIterator) Node() *enode.Node {
+	i.mutex.Lock()
+	defer i.mutex.Unlock()
+
+	if i.hasChangedToPreForkIterator {
+		return i.preForkIterator.Node()
+	}
+
+	return i.postForkIterator.Node()
+}
+
+// Check if the postForkIterator has a next element.
+// Once postForkIterator ends, switches to checking the preForkIterator.
+func (i *preAndPostForkIterator) Next() bool {
+	i.mutex.Lock()
+	defer i.mutex.Unlock()
+
+	if i.hasChangedToPreForkIterator {
+		return i.preForkIterator.Next()
+	}
+
+	hasNext := i.postForkIterator.Next()
+	if hasNext {
+		return true
+	}
+
+	i.hasChangedToPreForkIterator = true
+	return i.preForkIterator.Next()
+}
```

Comment:
we should probably Close the post-fork iterator here since it will never be used again

and in that case Close() should only close postFork if we didnt already switch to preFork (to not close twice)

Reply:
Agree, adding this